system/lib/wasmfs/file.h

```diff
@@ -283,7 +283,11 @@ class Directory::Handle : public File::Handle {
     if (name == "..") {
       return getParent();
     }
-    return getDir()->getChild(name);
+    auto child = getDir()->getChild(name);
+    if (child) {
+      child->locked().setParent(getDir());
+    }
+    return child;
```

Comment:
For the original in-memory backend, `getChild()` really did have no side effects because it could just return the child `File`. For persistent backends like node and opfs, though, the `File` does not exist before `getChild` is called, so the main purpose of `getChild` is to allocate and initialize a new `File` corresponding to a child that exists in the underlying storage API. That necessarily involves side effects.

The impetus for this PR was that I forgot to set the parent in the `getChild` implementation of the new opfs backend and it resulted in some confusing bugs. But the parent pointer is never used by backends; it's more like private metadata used only by the backend-independent logic. That makes it surprising that backends are responsible for initializing the parent pointer and this PR tries to eliminate that surprise. Moving the side effect of initializing the parent pointer to the target-independent logic makes the backend implementation less side-effectful, as you would expect.

Reply:
I can see how this PR is an improvement now, thanks. But I think my larger concern is that this breaks the invariant of `Directory::Handle::insertChild` being the only place that adds a child to a parent and updates both sides. It sounds like the node backend was doing that before this PR, which I wasn't aware of... That sounds like the cause of this part of your comment, I think:

> The impetus for this PR was that I forgot to set the parent in the getChild implementation of the new opfs backend and it resulted in some confusing bugs.

Overall, I'd hope we can have the node etc. backend create a new child when it wants to and then call `Directory::Handle::insertChild` to insert it, basically, and no other code would mess with parent pointers at all (except for the one method that removes a child, and also updates both sides atomically when doing so).

Reply:
`insertChild` also needs to be refactored to not take a `File` as an argument, but instead to tell the backend to allocate a `File` just like `getChild` does, with the difference being that the backend should also insert the new file into the underlying storage. Anyway, please don't depend on `insertChild` existing in its current form.

> Overall, I'd hope we can have the node etc. backend create a new child when it wants to and then call `Directory::Handle::insertChild` to insert it...

Why have each backend individually responsible for setting parent pointers, which they would otherwise be able to ignore entirely, when we could have backend-independent logic handle it correctly for all backends in a single location? This would make sense if backends controlled when new files were created, but they don't. Instead, they create files on demand when the backend-independent logic tells them to by calling the backend API. Having backends call out to the backend-independent logic would be an unnecessary inversion of control.

Reply:
I'm having trouble following you, sorry. Not sure why `insertChild` would need to be refactored. Maybe I wasn't clear enough, but more likely I'm missing something here... So this is what I'd hope we could write:
```diff
diff --git a/system/lib/wasmfs/backends/node_backend.cpp b/system/lib/wasmfs/backends/node_backend.cpp
index d7e356b69..f5393b01e 100644
--- a/system/lib/wasmfs/backends/node_backend.cpp
+++ b/system/lib/wasmfs/backends/node_backend.cpp
@@ -168,6 +168,7 @@ private:
     if (_wasmfs_node_get_mode(childPath.c_str(), &mode)) {
       return nullptr;
     }
+    // Create a new child, add it to ourselves, and return it.
     std::shared_ptr<File> child;
     if (S_ISREG(mode)) {
       child = std::make_shared<NodeFile>(mode, getBackend(), childPath);
@@ -179,7 +180,8 @@ private:
       // Unrecognized file kind not made visible to WasmFS.
       return nullptr;
     }
-    child->locked().setParent(shared_from_this()->cast<Directory>());
+    // Add the child.
+    locked().insertChild(name, child);
     return child;
   }
 ```
Is there a reason we can't?

Reply:
I'll upload a separate PR cleaning up `insertChild` that will clarify this discussion.